### PR TITLE
Add mobile UCRs for remote linked project spaces

### PR DESCRIPTION
**Pull request by proteusvacuum**
## Summary

https://dimagi-dev.atlassian.net/browse/SC-1474
I believe this should just work if the reports exist downstream, not sure why this was left out originally - @mkangia was this just because it was out of scope for what you were working on?

## Feature Flag

Linked project spaces, mobile ucr

## Safety Assurance

- [x] Risk label is set correctly
- [x] All migrations are backwards compatible and won't block deploy
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change
- [x] If QA is part of the safety story, the "Awaiting QA" label is used
- [x] I have confidence that this PR will not introduce a regression for the reasons below

### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

### CODE REVIEW

File: corehq/apps/app_manager/views/utils.py
Comment (mkangia) [inline corehq/apps/app_manager/views/utils.py:360]: I am not well aware of how linked UCRs work and my understanding was that this won't work for remote links because master_id won't be set.
If you say this would, i believe you.

Comment (proteusvacuum) [reply corehq/apps/app_manager/views/utils.py:360]: Yes, it gets set here:
https://github.com/dimagi/commcare-hq/blob/fr%2Fremote-linked-ucr/corehq/apps/linked_domain/ucr.py#L91

Comment (mkangia) [reply corehq/apps/app_manager/views/utils.py:360]: Thanks. All of this looks okay but I would still defer to someone who understands linked UCRs better to drop an approval.

Great to see that this solves issues with updating remotely linked apps as well!

File: corehq/apps/linked_domain/tests/test_linked_apps.py
Comment (snopoke) [inline corehq/apps/linked_domain/tests/test_linked_apps.py:171]: does this need to be saved? If the test fails the value won't get rolled back

**Review by snopoke — APPROVED** [review]
Regarding the tests, knowing how the code works I understand that testing the remote update functions with local data is OK however I do think it would be better to test it with data that isn't present in the local DB at all. This would prevent future changes from breaking the remote update.

I think this could be accomplished in one of two ways:

1. create unsaved versions of the data and manually set the `_id`
2. change the `_id` of the data returned from the mock to ensure it doesn't match anything in the DB

**Comment by proteusvacuum** [comment]
@snopoke ... so this actually doesn't work as expected, and the model of how this pushes to remote linked domains is backwards and pushing reports to remote linked domains doesn't work currently. I will fix this in this PR - marking as Don't Merge for now.

**Comment by mkangia** [comment]
Hey @proteusvacuum 

> the model of how this pushes to remote linked domains is backwards and pushing reports to remote linked domains doesn't work currently.

Can you explain this a bit more? I could not put this together.

**Comment by proteusvacuum** [comment]
@mkangia when you create a linked report, you do it from the "master" (upstream) domain and push to the "linked" (downstream) domain.
In the case of a remote linked domain, instead of pushing this to the remote server, we currently try and _fetch_ the report from the downstream server. 

For example, this function is called from the upstream server, where the user clicks "link reports"

https://github.com/dimagi/commcare-hq/blob/8dde8550c3afed44187d58a5a19a91145b99218a/corehq/apps/linked_domain/ucr.py#L22-L27

Line 24 tries to fetch the same report from the downstream server, and then carries on its merry way assuming there is a domain with the same name as the remotely linked domain (of the form `https://staging.commcarehq.org/a/linked-domain`). 

Luckily this actually fails here: https://sentry.io/organizations/dimagi/issues/2337546868, because that isn't a valid domain name. 

I will take full responsibility for this, since I'm the only one who has ever touched that code. I suppose since only a few projects use remote linked domains this hasn't actually come up yet. The only candidates for using this feature are eCHIS, and they only have static UCRs currently.

**Comment by mkangia** [comment]
thanks @proteusvacuum , I think I got it.

I always confuse terms upstream/downstream, so i look at the context of the explanation to understand which one is which.

if master is "upstream"

> In the case of a remote linked domain, instead of pushing this to the remote server, we currently try and fetch the report from the downstream server.

Did you mean fetch report from the upstream server here?

and

> For example, this function is called from the upstream server, where the user clicks "link reports"

downstream here?

My understanding from this was that the linked domain pulls from master in case of linked reports, and are currently failing due to incorrect domain name.

**Comment by snopoke** [comment]
I'm also a little confused here. I wasn't aware we ever 'push' change from upstream to downstream. I though we always pulled them from downstream.

**Comment by orangejenny** [comment]
@snopoke "Push" is a relatively new thing: we added it for USH last summer, because they have so many domains that it's arduous to pull each one individually. [docs](https://confluence.dimagi.com/display/ccinternal/Linked+Project+Spaces) (ctrl+F "push data")

**Comment by proteusvacuum** [comment]
@snopoke @mkangia , both SMS keywords and reports are linked by "pushing" them. Those features were built for the USH context. 
For keywords, we explicitly don't allow remote linked domains for that reason, but for reports I think my brain got twisted on how data was showing up where. Working on a spec now to allow "pulling" of reports from the downstream domain, which is the only way this will work for remote links (since the upstream server doesn't have any credentials on the downstream server).

**Comment by proteusvacuum** [comment]
https://github.com/dimagi/commcare-hq/pull/30018

**Pull request closed without merging**
